File: backend/app/dedupe.py
from __future__ import annotations

import re
import unicodedata
from datetime import date
from difflib import SequenceMatcher
from typing import Mapping


MERCHANT_NOISE = re.compile(r"股份有限公司|有限公司|企業社|分公司|門市部|門市|信用卡消費|簽帳消費|一般消費|消費款|交易")


def normalize_merchant(value: str) -> str:
    normalized = unicodedata.normalize("NFKC", value).upper()
    normalized = MERCHANT_NOISE.sub("", normalized)
    return "".join(char for char in normalized if char.isalnum() or "\u3400" <= char <= "\u9fff")


def merchant_similarity(left: str, right: str) -> float:
    a, b = normalize_merchant(left), normalize_merchant(right)
    if not a or not b:
        return 0.0
    if a == b:
        return 1.0
    if min(len(a), len(b)) >= 3 and (a in b or b in a):
        return 0.95
    return SequenceMatcher(None, a, b).ratio()


def is_duplicate(candidate: Mapping[str, object], existing: Mapping[str, object]) -> bool:
    if candidate["type"] != existing["type"]:
        return False
    if abs(float(candidate["amount"]) - float(existing["amount"])) > 0.01:
        return False
    days = abs((date.fromisoformat(str(candidate["date"])) - date.fromisoformat(str(existing["date"]))).days)
    similarity = merchant_similarity(str(candidate["title"]), str(existing["title"]))
    if days == 0 and similarity >= 0.9:
        return True
    carrier_pair = candidate["source"] == "雲端發票" or existing["source"] == "雲端發票"
    return bool(carrier_pair and days <= 3 and similarity >= 0.82)

File: lib/dedupe.ts
export type MatchableTransaction = {
  title: string;
  amount: number;
  date: string;
  type: "expense" | "income";
  source: string;
};

const merchantNoise = /(股份有限公司|有限公司|企業社|分公司|門市部|門市|信用卡消費|簽帳消費|一般消費|消費款|交易)/g;

export function normalizeMerchant(value: string) {
  return value
    .normalize("NFKC")
    .toUpperCase()
    .replace(merchantNoise, "")
    .replace(/[^0-9A-Z\u3400-\u9FFF]/g, "");
}

function dayDistance(a: string, b: string) {
  const left = Date.parse(`${a}T00:00:00Z`);
  const right = Date.parse(`${b}T00:00:00Z`);
  if (!Number.isFinite(left) || !Number.isFinite(right)) return Infinity;
  return Math.abs(left - right) / 86_400_000;
}

function bigrams(value: string) {
  if (value.length < 2) return [value];
  return Array.from({ length: value.length - 1 }, (_, index) => value.slice(index, index + 2));
}

export function merchantSimilarity(a: string, b: string) {
  const left = normalizeMerchant(a);
  const right = normalizeMerchant(b);
  if (!left || !right) return 0;
  if (left === right) return 1;
  if (Math.min(left.length, right.length) >= 3 && (left.includes(right) || right.includes(left))) return 0.95;
  const available = bigrams(right);
  let matches = 0;
  for (const gram of bigrams(left)) {
    const index = available.indexOf(gram);
    if (index >= 0) { matches += 1; available.splice(index, 1); }
  }
  return (2 * matches) / (bigrams(left).length + bigrams(right).length);
}

export function isDuplicateTransaction(candidate: MatchableTransaction, existing: MatchableTransaction) {
  if (candidate.type !== existing.type || Math.abs(candidate.amount - existing.amount) > 0.01) return false;

  const days = dayDistance(candidate.date, existing.date);
  const similarity = merchantSimilarity(candidate.title, existing.title);

  // 同日、同金額、同店家的重複匯入。
  if (days === 0 && similarity >= 0.9) return true;

  // 銀行入帳日可能比載具發票晚；只對雲端發票放寬到三天。
  const carrierPair = candidate.source === "雲端發票" || existing.source === "雲端發票";
  return carrierPair && days <= 3 && similarity >= 0.82;
}
